src/rdkafka_telemetry.c

```diff
@@ -0,0 +1,785 @@
+                        compression_used = RD_KAFKA_COMPRESSION_ZSTD;
+                        break;
+#endif
+#if WITH_SNAPPY
+                case RD_KAFKA_COMPRESSION_SNAPPY:
+                        r = rd_kafka_compress_snappy(rkb, payload, payload_len,
+                                                     compressed_payload,
+                                                     compressed_payload_size);
+                        compression_used = RD_KAFKA_COMPRESSION_SNAPPY;
+                        break;
+#endif
+                default:
+                        break;
+                }
+                if (compression_used != RD_KAFKA_COMPRESSION_NONE && r == 0) {
+                        rd_kafka_dbg(
+                            rk, TELEMETRY, "PUSH",
+                            "Compressed payload of size %" PRIusz " to %" PRIusz
+                            " using compression type "
+                            "%s",
+                            payload_len, *compressed_payload_size,
+                            rd_kafka_compression2str(compression_used));
+                        rd_free(payload);
+                        return compression_used;
+                }
+        }
+        if (compression_used != RD_KAFKA_COMPRESSION_NONE && r == -1) {
+                rd_kafka_dbg(rk, TELEMETRY, "PUSHERR",
+                             "Failed to compress payload with available "
+                             "compression types");
+        }
+        rd_kafka_dbg(rk, TELEMETRY, "PUSH", "Sending uncompressed payload");
+
+        *compressed_payload      = payload;
+        *compressed_payload_size = payload_len;
+        return RD_KAFKA_COMPRESSION_NONE;
+}
+
+
+static void rd_kafka_send_push_telemetry(rd_kafka_t *rk,
+                                         rd_kafka_broker_t *rkb,
+                                         rd_bool_t terminating) {
+
+        size_t metrics_payload_size = 0, compressed_metrics_payload_size = 0;
+        void *metrics_payload =
+                 rd_kafka_telemetry_encode_metrics(rk, &metrics_payload_size),
+             *compressed_metrics_payload        = NULL;
+        rd_kafka_compression_t compression_used = RD_KAFKA_COMPRESSION_NONE;
+
+        if (rk->rk_telemetry.accepted_compression_types_cnt != 0) {
+                compression_used = rd_kafka_push_telemetry_payload_compress(
+                    rk, rkb, metrics_payload, metrics_payload_size,
+                    &compressed_metrics_payload,
+                    &compressed_metrics_payload_size);
+        } else {
+                rd_kafka_dbg(rk, TELEMETRY, "PUSHSENT",
+                             "No compression types accepted, sending "
+                             "uncompressed payload");
+                compressed_metrics_payload      = metrics_payload;
+                metrics_payload                 = NULL;
+                compressed_metrics_payload_size = metrics_payload_size;
+        }
```

Comment:
We can pass the `rd_buf_t` for compression here.
```suggestion
        rd_buf_t *metrics_payload = rd_kafka_telemetry_encode_metrics(rk);
        size_t compressed_metrics_payload_size = 0;
        void *compressed_metrics_payload       = NULL;
        rd_kafka_compression_t compression_used =
            rd_kafka_push_telemetry_payload_compress(
                rk, rkb, metrics_payload, &compressed_metrics_payload,
                &compressed_metrics_payload_size);
        if (compressed_metrics_payload_size >
            (size_t)rk->rk_telemetry.telemetry_max_bytes) {
                rd_kafka_log(rk, LOG_WARNING, "TELEMETRY",
                             "Metrics payload size %" PRIdsz
                             " exceeds telemetry_max_bytes %" PRId32
                             "specified by the broker.",
                             compressed_metrics_payload_size,
                             rk->rk_telemetry.telemetry_max_bytes);
        }
```

Reply:
Updated